### PR TITLE
types: fix json_unquote

### LINKED ISSUE #25546: json_unquote doesn't detect invalid json string
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

    SELECT  json_unquote('""a""');
<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

    mysql> SELECT  json_unquote('""a""');
    ERROR 3141 (22032): Invalid JSON text in argument 1 to function json_unquote: "The document root must not be followed by other values." at position 2.
### 3. What did you see instead (Required)

    mysql> SELECT  json_unquote('""a""');
    +-----------------------+
    | json_unquote('""a""') |
    +-----------------------+
    | "a"                   |
    +-----------------------+
    1 row in set (0.01 sec)
### 4. What is your TiDB version? (Required)

    mysql> select tidb_version();
    +--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
    | tidb_version()                                                                                                                                                                                                                                                                                                           |
    +--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
    | Release Version: v5.2.0-alpha-94-g9e08fc559
    Edition: Community
    Git Commit Hash: 9e08fc559c8175d00aefd71fc84667ac4d5dfc21
    Git Branch: master
    UTC Build Time: 2021-06-18 02:14:53
    GoVersion: go1.16.4
    Race Enabled: false
    TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
    Check Table Before Drop: false |
    +--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
<!-- Paste the output of SELECT tidb_version() -->

### PR DESCRIPTION
**Pull request by wzru**
### What problem does this PR solve?

Issue Number: close #25546 

Problem Summary:

json_unquote() bug when input is like `""a""`
expected output:
```bash
mysql> SELECT  json_unquote('""a""');
ERROR 3141 (22032): Invalid JSON text in argument 1 to function json_unquote: "The document root must not be followed by other values." at position 2.
```
wrong output:
```
mysql> SELECT  json_unquote('""a""');
+-----------------------+
| json_unquote('""a""') |
+-----------------------+
| "a"                   |
+-----------------------+
1 row in set (0.01 sec)
```

### What is changed and how it works?

What's Changed:

BinaryJSON.Unquote() implementation

How it Works:

Check if json is valid when unquote

### Check List 

Tests 

- Unit test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM

### Release note 

- No release note

### CODE REVIEW

File: expression/builtin_json.go [hunk 1]
```diff
@@ -216,11 +216,14 @@ func (c *jsonUnquoteFunctionClass) getFunction(ctx sessionctx.Context, args []Ex
 	return sig, nil
 }
 
-func (b *builtinJSONUnquoteSig) evalString(row chunk.Row) (string, bool, error) {
+func (b *builtinJSONUnquoteSig) evalString(row chunk.Row) (res string, isNull bool, err error) {
```
Comment (guo-shaoge) [inline expression/builtin_json.go:219]: res -> str

File: expression/builtin_json.go [hunk 2]
```diff
@@ -216,11 +216,14 @@ func (c *jsonUnquoteFunctionClass) getFunction(ctx sessionctx.Context, args []Ex
...
 	str, isNull, err := b.args[0].EvalString(b.ctx, row)
 	if isNull || err != nil {
 		return "", isNull, err
 	}
+	if len(str) >= 2 && str[0] == '"' && str[len(str)-1] == '"' && !goJSON.Valid([]byte(str)) {
+		return "", false, json.ErrInvalidJSONText.GenWithStackByArgs("The document root must not be followed by other values.")
```
Comment (guo-shaoge) [inline expression/builtin_json.go:225]: Looks like still not compatible with MySQL. Maybe should add an integrated test.

    drop table if exists t1;
    create table t1(c1 varchar(100));
    insert into t1 values('""abc""');
    insert into t1 values('""abc""');
    insert into t1 values('""abc""');
    select json_unquote(c1) from t1;
    mysql> select json_unquote(c1) from t1;
    +------------------+
    | json_unquote(c1) |
    +------------------+
    | "abc"            |
    | "abc"            |
    | "abc"            |
    +------------------+
    3 rows in set (0.01 sec)

File: expression/builtin_json.go (hunk 1 above)
Comment (wzru) [reply expression/builtin_json.go]: modified

File: expression/builtin_json.go (hunk 2 above)
Comment (wzru) [reply expression/builtin_json.go:225]: fixed

Comment (guo-shaoge) [reply expression/builtin_json.go:225]: OK. Please complete the corresponding modification of COPR in time. Otherwise the behavior will not be compatible.

**Comment by guo-shaoge** [comment]
Maybe no need to update master too soon. The bot will update automatically when it's ready for merging.

**Review by guo-shaoge — APPROVED** [review]

**Review by ichn-hu — APPROVED** [review]

**Comment by skyzh** [comment]
I have already raised this issue in TiKV Coprocessor months ago, https://github.com/tikv/tikv/pull/10177#issuecomment-850263100

We need to fix this in copr too......

**Comment by guo-shaoge** [comment]
> I have already raised this issue in TiKV Coprocessor months ago, [tikv/tikv#10177 (comment)](https://github.com/tikv/tikv/pull/10177#issuecomment-850263100)
> 
> We need to fix this in copr too......
> 
> > Also I found this logic might be inconsistent with MySQL if the JSON is invalid. Consider this case:
> 
> > ```
> > SELECT  json_unquote('""a""');
> > ```
> 
> > where `""a""` is an invalid JSON object, and `json_unquote` might yield `"a"` after all. I think this might be a bug that should be fixed in both TiDB and TiKV in the future.

Guess @wzru  is working on it.

**Pull request merged**